### PR TITLE
Fix bug in swank/backend:compiler-macroexpand-1

### PR DESCRIPTION
**Pull request by gdobbins**
This fixes issue #339

### CODE REVIEW

File: swank/backend.lisp [hunk 1]
```diff
@@ -719,7 +719,7 @@ return the results and T.  Otherwise, return the original form and
 NIL."
   (let ((fun (and (consp form)
                   (valid-function-name-p (car form))
-                  (compiler-macro-function (car form)))))
+                  (compiler-macro-function (car form) env))))
```
Comment (luismbo) [inline swank/backend.lisp]: How does this fix work? swank-macrostep is not passing an environment to `compiler-macroexpand-1` is it?

Comment (gdobbins) [inline swank/backend.lisp]: Running the example in issue #339 without this fix on SBCL 1.3.9 gives the following:

    error while parsing arguments to DEFINE-COMPILER-MACRO FOO:
      too many elements in
        (1 2 3)
      to satisfy lambda list
        (X Y):
      exactly 2 expected, but got 3
       [Condition of type SB-KERNEL::ARG-COUNT-ERROR]
    
    Restarts:
     0: [*ABORT] Return to SLIME's top level.
     1: [ABORT] abort thread (#<THREAD "repl-thread" RUNNING {1005090033}>)
    
    Backtrace:
      0: (SB-C::CHECK-DS-LIST #<unavailable argument> #<unavailable argument> #<unavailable argument> #<unavailable argument>)
      1: ((COMPILER-MACRO FOO) (FOO 1 2 3) #<unavailable argument>)
      2: ((LAMBDA (SWANK/BACKEND::FORM &OPTIONAL SWANK/BACKEND::ENV) :IN "/home/graham/.emacs.d/elpa/slime-20160907.602/swank/backend.lisp") (FOO 1 2 3) #S(SB-KERNEL:LEXENV :FUNS ((FOO . #) (#1=#:G4 SB-SYS:MAC..
      3: (SWANK/BACKEND:COMPILER-MACROEXPAND-1 (FOO 1 2 3) #S(SB-KERNEL:LEXENV :FUNS ((FOO . #) (#1=#:G4 SB-SYS:MACRO . #<CLOSURE #2=# {100729CDDB}>) (#1# SB-SYS:MACRO . #<CLOSURE #2# {100729CC8B}>)) :VARS NIL..
      4: ((LAMBDA (SWANK/SBCL::FORM SWANK/SBCL::CONTEXT SWANK/SBCL::ENVIRONMENT) :IN SWANK/BACKEND:COLLECT-MACRO-FORMS) (FOO 1 2 3) #<unused argument> #S(SB-KERNEL:LEXENV :FUNS ((FOO . #) (#1=#:G4 SB-SYS:MACRO..
      5: (SB-WALKER::WALK-FORM-INTERNAL (FOO 1 2 3) :EVAL #S(SB-KERNEL:LEXENV :FUNS ((FOO . #) (#1=#:G4 SB-SYS:MACRO . #<CLOSURE #2=# {100729CDDB}>) (#1# SB-SYS:MACRO . #<CLOSURE #2# {100729CC8B}>)) :VARS NIL ..
      6: (SB-WALKER::WALK-REPEAT-EVAL ((FOO 1 2 3)) #S(SB-KERNEL:LEXENV :FUNS ((FOO . #) (#1=#:G4 SB-SYS:MACRO . #<CLOSURE #2=# {100729CDDB}>) (#1# SB-SYS:MACRO . #<CLOSURE #2# {100729CC8B}>)) :VARS NIL :BLOCK..
      7: (SB-WALKER::WALK-FLET (FLET ((FOO # ARGS)) (FOO 1 2 3)) :EVAL #S(SB-KERNEL:LEXENV :FUNS ((#:G4 SB-SYS:MACRO . #<CLOSURE # {100729CC8B}>)) :VARS NIL :BLOCKS NIL :TAGS NIL :TYPE-RESTRICTIONS NIL :LAMBDA..
      8: (SB-WALKER::WALK-FORM-INTERNAL (FLET ((FOO # ARGS)) (FOO 1 2 3)) :EVAL #S(SB-KERNEL:LEXENV :FUNS ((#:G4 SB-SYS:MACRO . #<CLOSURE # {100729CC8B}>)) :VARS NIL :BLOCKS NIL :TAGS NIL :TYPE-RESTRICTIONS NI..
      9: ((FLET SWANK/BACKEND:COLLECT-MACRO-FORMS :IN "/home/graham/.emacs.d/elpa/slime-20160907.602/swank/sbcl.lisp") (FLET ((FOO # ARGS)) (FOO 1 2 3)) NIL)
     10: (SWANK-MACROSTEP::COLLECT-MACRO-FORMS-IN-CONTEXT (FLET ((FOO # ARGS)) (FOO 1 2 3)) ("" ""))
     11: ((LAMBDA NIL :IN SWANK-MACROSTEP:MACROSTEP-EXPAND-1))
     12: (SWANK::CALL-WITH-BUFFER-SYNTAX NIL #<CLOSURE (LAMBDA NIL :IN SWANK-MACROSTEP:MACROSTEP-EXPAND-1) {100729AAEB}>)
     13: (SB-INT:SIMPLE-EVAL-IN-LEXENV (SWANK-MACROSTEP:MACROSTEP-EXPAND-1 "(bar (foo 1 2 3))" T (QUOTE ("" ""))) #<NULL-LEXENV>)
     14: (EVAL (SWANK-MACROSTEP:MACROSTEP-EXPAND-1 "(bar (foo 1 2 3))" T (QUOTE ("" ""))))
     15: (SWANK:EVAL-FOR-EMACS (SWANK-MACROSTEP:MACROSTEP-EXPAND-1 "(bar (foo 1 2 3))" T (QUOTE ("" ""))) "COMMON-LISP-USER" 54)
     16: (SWANK::PROCESS-REQUESTS NIL)
     17: ((LAMBDA NIL :IN SWANK::HANDLE-REQUESTS))
     18: ((LAMBDA NIL :IN SWANK::HANDLE-REQUESTS))
     19: (SWANK/SBCL::CALL-WITH-BREAK-HOOK #<FUNCTION SWANK:SWANK-DEBUGGER-HOOK> #<CLOSURE (LAMBDA NIL :IN SWANK::HANDLE-REQUESTS) {100509800B}>)
     20: ((FLET SWANK/BACKEND:CALL-WITH-DEBUGGER-HOOK :IN "/home/graham/.emacs.d/elpa/slime-20160907.602/swank/sbcl.lisp") #<FUNCTION SWANK:SWANK-DEBUGGER-HOOK> #<CLOSURE (LAMBDA NIL :IN SWANK::HANDLE-REQUESTS..
     21: (SWANK::CALL-WITH-BINDINGS ((*STANDARD-INPUT* . #1=#<SWANK/GRAY::SLIME-INPUT-STREAM {1004FA5A63}>) (*STANDARD-OUTPUT* . #2=#<SWANK/GRAY::SLIME-OUTPUT-STREAM {1005077053}>) (*TRACE-OUTPUT* . #2#) (*ERR..
     22: (SWANK::HANDLE-REQUESTS #<SWANK::MULTITHREADED-CONNECTION {10045DC8E3}> NIL)
     23: ((FLET #:WITHOUT-INTERRUPTS-BODY-1158 :IN SB-THREAD::INITIAL-THREAD-FUNCTION-TRAMPOLINE))
     24: ((FLET SB-THREAD::WITH-MUTEX-THUNK :IN SB-THREAD::INITIAL-THREAD-FUNCTION-TRAMPOLINE))
     25: ((FLET #:WITHOUT-INTERRUPTS-BODY-359 :IN SB-THREAD::CALL-WITH-MUTEX))
     26: (SB-THREAD::CALL-WITH-MUTEX #<CLOSURE (FLET SB-THREAD::WITH-MUTEX-THUNK :IN SB-THREAD::INITIAL-THREAD-FUNCTION-TRAMPOLINE) {7FFFE890ED6B}> #<SB-THREAD:MUTEX "thread result lock" owner: #<SB-THREAD:THR..
     27: (SB-THREAD::INITIAL-THREAD-FUNCTION-TRAMPOLINE #<SB-THREAD:THREAD "repl-thread" RUNNING {1005090033}> NIL #<CLOSURE (LAMBDA NIL :IN SWANK-REPL::SPAWN-REPL-THREAD) {100508FF9B}> (#<SB-THREAD:THREAD "re..
     28: ("foreign function: call_into_lisp")
     29: ("foreign function: new_thread_trampoline")
    
Looking at position 3 in the backtrace shows that `swank/backend:compiler-macroexpand-1` is in fact being called with an environment. Inspecting the environment shows

    (FOO . #<SB-C::FUNCTIONAL :%SOURCE-NAME SB-C::.ANONYMOUS. :%DEBUG-NAME NIL {100729D073}>)

among the functions, which is correct as `bar` defines `foo` via flet. But `compiler-macroexpand-1` is incorrectly not using this passed environment for its expansion of compiler macros.

**Comment by gdobbins** [comment]
Any word on this?

**Comment by attila-lendvai** [comment]
i could reproduce this and the fix indeed fixes it.

**Pull request merged**